lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,221 @@
+          Fog::Logger.warning("Template processed. Populated #{@files}")
+          @files
+        end
+
+        def files_basepath
+          min_lenght = @files.keys.map(&:length).min
+          candidates = @files.keys.map { |x| File.dirname(x[0..min_lenght]) }.to_set
+          return nil if candidates.size != 1
+          candidates.each do |x|
+            return x
+          end
+        end
+
+        def is_template(content)
+          # Return true if the file is an heat template, false otherwise.
+          htv = content.index("heat_template_version:")
+          !!(htv && htv < 5)
+        end
+
+        def fixup_uri(uri)
+          # Ruby URI is not round-trip safe when schema == file
+          #  eg. URI("file:///a.out").to_s != file:///a.out"
+          if uri.kind_of?(URI) && uri.scheme == "file"
+            uri.host = ""
+          end
+          uri
+        end
+
+        def ignore_if(key, value)
+          # Should I attach this file?
+          return true if key != 'get_file' && key != 'type'
+
+          return true unless value.kind_of?(String)
+
+          return true if key == 'type' &&
+                         !value.end_with?('.yaml', '.template')
+
+          false
+        end
+
+        def recurse_if(value)
+          # Should I recurse into this template branch?
+          !!(value.kind_of?(Hash) || value.kind_of?(Array))
+        end
+
+        def url_join(prefix, suffix)
+          # return string
+          if prefix
+            suffix = URI.join(prefix, suffix)
+            fixup_uri(suffix)
+            suffix = suffix.to_s
+          end
+          suffix
+        end
+
+        def base_url_for_url(url)
+          # Returns the string baseurl of the given url.
+          parsed = URI(url)
+          # Ruby URI is not round-trip safe when schema == file
+          fixup_uri(parsed)
+          parsed_dir = File.dirname(parsed.path)
+          URI.join(parsed, parsed_dir).to_s
+        end
+
+        def normalise_file_path_to_url(path)
+          # Nothing to do on URIs
+          return path if URI(path).scheme
+
+          path = File.absolute_path(path)
+          url_join('file:', path)
+        end
+
+        def get_content(uri_or_filename)
+          Fog::Logger.warning("Opening #{uri_or_filename}")
+          uri_or_filename = fixup_uri(uri_or_filename)
+          uri_or_filename = uri_or_filename.to_s if uri_or_filename.kind_of?(URI)
+          # throw exceptions enables stack creation to fail
+          #   with a suitable error.
+          #
+          # XXX Implement a retry here?
+          content = ''
+          # does ruby close the socket/filedescriptor at exit?
+          # XXX Limit download file size
+          # XXX Protect from vanilla open-uri attacks
+          uri_or_filename = uri_or_filename[7..-1] if uri_or_filename.start_with?("file:///")
+          open(uri_or_filename) { |f| content = f.read }
+          content == "Error" ? nil : content
```

Comment:
This line is problematic. You are not assigning the result to anything. Besides, can `content` ever be `Error`?
You made a few comments but did not provide solution. It does not look like final code to be merged.

Reply:
I've been told to create a  [WIP] pull request to simplify code reviews.

For these lines I'd appreciate some help: 

  - which are some common test cases involving open-uri?
  - do you have any suggestions about securing other parts of this code?